### DISCUSSION TITLE
$PATH set, old version still used

### DISCUSSION THREAD

**Discussion by alan-taylor** [Homebrew/discussions#565] [Everyday usage]
Hello,
I have just installed homebrew on a MacBook running macOS 11.1, apparently successfully. Then install a new version of rsync using homebrew. Following that, the system still uses the old original system version of rsync, despite the $PATH being set correctly ?

$ xcode-select --install
xcode-select: error: command line tools are already installed, use "Software Update" to install updates

$ /bin/bash install.sh
lots of seemingly normal output, this is a selection near the end ...
 * [new branch]      master     -> origin/master
Checking out files: 100% (5737/5737), done.
HEAD is now at 3bf9f8cc9d kibana: deprecate (#69122)
==> Installation successful!

The $PATH seems already set correctly by default, I haven't changed anything:
% echo $PATH
/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
% cat /private/etc/paths
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
% 

I installed rsysnc using homebrew, apparently successfully:
% which rsync
/usr/bin/rsync
% which brew
/usr/local/bin/brew
% ls -lL /usr/local/bin/rsync
-r-xr-xr-x  1 alan  staff  480608 16 Jan 20:09 /usr/local/bin/rsync
% otool -L /usr/local/bin/rsync
/usr/local/bin/rsync:
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/local/opt/popt/lib/libpopt.0.dylib (compatibility version 1.0.0, current version 1.1.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/local/opt/lz4/lib/liblz4.1.dylib (compatibility version 1.0.0, current version 1.9.2)
	/usr/local/opt/zstd/lib/libzstd.1.dylib (compatibility version 1.0.0, current version 1.4.5)
	/usr/local/opt/xxhash/lib/libxxhash.0.dylib (compatibility version 0.0.0, current version 0.8.0)
	/usr/local/opt/openssl@1.1/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.0.0)
	/usr/lib/libcharset.1.dylib (compatibility version 2.0.0, current version 2.0.0)

However, it is the old system rsync which is still being used unless I specify the complete path:
% rsync --version
rsync  version 2.6.9  protocol version 29
... etc
% /usr/local/bin/rsync --version
rsync  version 3.2.3  protocol version 31
... etc
%

Any thoughts on further trouble shooting ?

BRgds/Alan

**Comment by gromgit** [comment]
Hmm, your shell prompt suddenly went from `$` to `%`, but I'll assume the following:
* you're still running in the same shell session in which you `brew install rsync`
* you're running a bash shell session (everything I mention should still be true for zsh, but I might be wrong)
* running `type -p rsync` outputs `/usr/bin/rsync`

That last point means you ran `rsync` in this session before you installed the Homebrew version, so that (old) path is _cached_ in bash's command lookup table to avoid having to search for it again.

Run `hash -r` to empty that table, and now running `rsync` will do a normal `PATH` traversal to find the command...and cache the path it finds.

**Comment by alan-taylor** [comment]
SOLVED !
Thank you :)
I did not know bash (and apparently zsh as that is the shell) had that cache behaviour.

As you suggested:
% hash -r
% rsync --version               
rsync  version 3.2.3  protocol version 31
etc ...

Problem solved, many thanks.
BRgds/Alan
